lib/charon/session_store/redis_store.ex

```diff
@@ -46,101 +46,109 @@ defmodule Charon.SessionStore.RedisStore do
 
   @impl true
+  def upsert(_session = %{refresh_expires_at: exp, refreshed_at: now}, _config) when exp < now,
+    do: :ok
+
   def upsert(
         session = %{
           id: sid,
           user_id: uid,
           type: type,
           refreshed_at: now,
-          expires_at: s_exp,
           refresh_expires_at: exp
         },
         config
       ) do
     mod_conf = get_mod_config(config)
-    {uid_str, type_str, key_prefix} = key_data(uid, type, mod_conf)
-    session_key = session_key(sid, uid_str, type_str, key_prefix)
-    set_key = set_key(uid_str, type_str, key_prefix)
+    key_data = key_data(uid, type, mod_conf)
+    session_set_key = session_set_key(key_data)
+    exp_oset_key = exp_oset_key(key_data)
     exp_str = Integer.to_string(exp)
     now = Integer.to_string(now)
 
-    # upsert the actual session as a separate key-value pair that expires when the refresh token expires
-    upsert_session_c = ["SET", session_key, serialize(session, mod_conf, config), "EXAT", exp_str]
-    # add session key to user's session set, with exp timestamp as score (or update score)
-    upsert_set_c = ["ZADD", set_key, exp_str, session_key]
-    max_exp_c = get_max_exp_session_cmd(set_key)
-    prune_set_c = prune_session_set_cmd(set_key, now)
-
-    if s_exp == exp do
-      # s_exp == exp, we can't assume that this session has the highest refresh exp
-      # because this session's refresh_expires_at is reduced so that it is <= expires_at
-
-      [@multi, max_exp_c, upsert_session_c, upsert_set_c, prune_set_c, @exec]
-      |> mod_conf.redix_module.pipeline()
-      |> case do
-        {:ok, [_, _, _, _, _, [prev_max_exp_session, _, _, _]]} ->
-          # prev_max_exp_session had the highest exp *before* this session was upserted
-          # update user session set ttl if there is no other session OR the new session's exp is the highest
-          {prev_exists?, prev_max_exp_str} = parse_zrange_withscores(prev_max_exp_session)
-          prev_max_exp = String.to_integer(prev_max_exp_str)
-          if not prev_exists? or exp > prev_max_exp, do: put_s_exp(set_key, exp_str, mod_conf)
-          :ok
-
-        error ->
-          redis_result_to_error(error)
-      end
-    else
-      # s_exp != exp; we can assume that this session has the highest refresh exp of all of the user's sessions
-      set_exp_c = put_s_exp_cmd(set_key, exp_str)
-
-      [@multi, upsert_session_c, upsert_set_c, set_exp_c, prune_set_c, @exec]
-      |> mod_conf.redix_module.pipeline()
-      |> case do
-        {:ok, [_, _, _, _, _, [_, _, _, _]]} -> :ok
-        error -> redis_result_to_error(error)
-      end
+    serialized = serialize(session, mod_conf, config)
+    new_upsert_c = ["HSET", session_set_key, sid, serialized]
+    exp_new_upsert_c = ["EXPIRE", session_set_key, to_string(config.refresh_token_ttl)]
+    new_exp_oset_c = ["ZADD", exp_oset_key, exp_str, sid]
+    exp_new_exp_oset_c = ["EXPIRE", exp_oset_key, to_string(config.refresh_token_ttl)]
```

Comment:
to simplify things, we expire the user's session sets at `refresh_token_ttl`, regardless of the actual exp of the current sessions compared to other sessions in the set. In v3, this is properly taken into account again.